core/trino-main/src/main/java/io/trino/execution/executor/MultilevelSplitQueue.java

```diff
@@ -304,6 +313,18 @@ public static int computeLevel(long threadUsageNanos)
         return LEVEL_THRESHOLD_SECONDS.length - 1;
     }
 
+    public DistributionStat getWaitingSplitsSizeDistribution()
+    {
+        return waitingSplitsSizeDistribution;
+    }
+
+    private void recordWaitingSplitQueueSize()
+    {
+        long now = ticker.read();
+        this.waitingSplitsSizeDistribution.add(this.size(), now - this.lastTimeSplitSizeWasRecorded);
```

Comment:
Ticker uses `System.nanoTIme()` that does not have nanosecond resolution (I think it's about a microsecond).
So the first issue here is that `now - this.lastTimeSplitSizeWasRecorded` can be 0, I'm not sure how it influences the histogram, we may add `min(1, diff) `here to fix that.

The other issue is that you can have the diff 0 or 1000 for very similar values of tick. This can influence how the distribution will look like, not sure if it will in practice.

Reply:
> Ticker uses System.nanoTIme() that does not have nanosecond resolution (I think it's about a microsecond).
So the first issue here is that now - this.lastTimeSplitSizeWasRecorded can be 0, I'm not sure how it influences the histogram, we may add min(1, diff) here to fix that.

Yes, it is possible but it is a kind of cheating. If we observe two same ticks for two calls it most probably means that the period between them was very short. It means that the observed value was very short-living.
If we normalize it to the `1` we lie, if we do not, it disappears (becase the weight is `0`). Not sure what is better here. Most probably we just to [nanotrust here in nanoTime](https://shipilev.net/blog/2014/nanotrusting-nanotime/) ;).

> The other issue is that you can have the diff 0 or 1000 for very similar values of tick. This can influence how the distribution will look like, not sure if it will in practice.

I do not think that there is a better way to do it without using `System.nanoTime`. It is basing on CPU timer and there is no much guarantees, that's true but what we can do?